src/main/java/com/yelp/nrtsearch/server/luceneserver/search/query/multifunction/GeoPointDecayFilterFunction.java

```diff
@@ -0,0 +1,179 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.yelp.nrtsearch.server.luceneserver.search.query.multifunction;
+
+import com.google.type.LatLng;
+import com.yelp.nrtsearch.server.grpc.MultiFunctionScoreQuery;
+import com.yelp.nrtsearch.server.luceneserver.IndexState;
+import com.yelp.nrtsearch.server.luceneserver.doc.LoadedDocValues;
+import com.yelp.nrtsearch.server.luceneserver.doc.SegmentDocLookup;
+import com.yelp.nrtsearch.server.luceneserver.field.LatLonFieldDef;
+import com.yelp.nrtsearch.server.luceneserver.geo.GeoPoint;
+import com.yelp.nrtsearch.server.luceneserver.geo.GeoUtils;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import org.apache.lucene.index.*;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.Query;
+
+public class GeoPointDecayFilterFunction extends DecayFilterFunction {
+
+  private final MultiFunctionScoreQuery.DecayFunction decayFunction;
+  private final String fieldName;
+  private final DecayFunction decayType;
+  private final double scale;
+  private final double offset;
+  private final double decay;
+  private final LatLng origin;
+  private final IndexState indexState;
+
+  /**
+   * Constructor.
+   *
+   * @param filterQuery filter to use when applying this function, or null if none
+   * @param weight weight multiple to scale the function score
+   * @param decayFunction to score a document with a function that decays depending on the distance
+   *     between an origin point and a numeric doc field value
+   * @param indexState indexState for validation and doc value lookup
+   */
+  public GeoPointDecayFilterFunction(
+      Query filterQuery,
+      float weight,
+      MultiFunctionScoreQuery.DecayFunction decayFunction,
+      IndexState indexState) {
+    super(filterQuery, weight, decayFunction);
+    this.decayFunction = decayFunction;
+    this.fieldName = decayFunction.getFieldName();
+    this.decayType = getDecayType(decayFunction.getDecayType());
+    this.offset =
+        !decayFunction.getOffset().isEmpty()
+            ? GeoUtils.getDistance(decayFunction.getOffset())
+            : 0.0;
+    this.decay = decayFunction.getDecay();
+    double userGivenScale = GeoUtils.getDistance(decayFunction.getScale());
+    this.scale = decayType.computeScale(userGivenScale, decay);
+    this.indexState = indexState;
+    if (!decayFunction.hasGeoPoint()) {
+      throw new IllegalArgumentException("Decay Function should have a geoPoint for Origin field");
+    } else {
+      this.origin = decayFunction.getGeoPoint();
+    }
+    LatLonFieldDef latLonFieldDef = (LatLonFieldDef) indexState.getField(fieldName);
+    // TODO: Add support for multi-value fields
+    if (latLonFieldDef.isMultiValue()) {
+      throw new IllegalArgumentException(
+          "Multivalue fields are not supported for decay functions yet");
+    }
```

Comment:
It would be good to check that the fields is actually the correct type, and throw a descriptive exception if it is not. It would also be good to verify the field has doc values enabled.